File: src/main/java/com/ra/base_spring_boot/model/Movie.java
package com.ra.base_spring_boot.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.ra.base_spring_boot.model.base.BaseObject;
import com.ra.base_spring_boot.model.constants.MovieType;
import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import lombok.*;

import java.util.Date;
import java.util.Set;

@Entity
@Table(name = "movies")
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Builder
public class Movie extends BaseObject {

    @NotBlank(message = "Movie name cannot be blank")
    @Column(name = "title", nullable = false, length = 255)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "author", length = 100, nullable = false)
    private String author;

    @Column(name = "image", length = 255)
    private String image;

    @Column(name = "trailer", length = 255)
    private String trailer;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MovieType type;

    @Min(value = 1, message = "Duration must be greater than 0 minutes")
    @Column(nullable = false)
    private int duration;

    @Column(name = "release_date")
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX", timezone = "Asia/Ho_Chi_Minh")
    private Date releaseDate;

    @Column(name = "actors", columnDefinition = "TEXT")
    private String actors;

    @Column(name = "nation", length = 100)
    private String nation;

    @Column(name = "created_at")
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX", timezone = "Asia/Ho_Chi_Minh")
    private Date createdAt;

    @Column(name = "updated_at")
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX", timezone = "Asia/Ho_Chi_Minh")
    private Date updatedAt;

    @ManyToMany
    @JoinTable(
            name = "movie_genre",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "genre_id")
    )
    private Set<Genre> genres;

    @PrePersist
    protected void onCreate() {
        this.createdAt = new Date();
        this.updatedAt = new Date();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = new Date();
    }
}

File: api/src/models/movie.js
// models/movie.model.js
const mongoose = require('mongoose');

const movieSchema = new mongoose.Schema({
  title: { type: String, required: true, maxlength: 255 },
  description: { type: String }, // Fixed: description not descriptions
  author: { type: String, required: true, maxlength: 100 },
  image: { type: String, maxlength: 255 },
  trailer: { type: String, maxlength: 255 },
  type: { type: String, enum: ['2D', '3D'], required: true },
  duration: { type: Number, min: 1, required: true },
  release_date: { type: Date },
  actors: { type: String }, // Added missing field
  nation: { type: String, maxlength: 100 }, // Added missing field
  genres: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Genre' }],
}, { timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' } });

module.exports = mongoose.model('Movie', movieSchema);
